sdv/constraints/tabular.py

```diff
@@ -61,70 +77,95 @@ class CustomConstraint(Constraint):
             Function to replace the ``reverse_transform`` method.
         is_valid (callable):
             Function to replace the ``is_valid`` method.
+
+    Returns:
+        CustomConstraint class:
+            A constraint with custom ``transform``/``reverse_transform``/``is_valid`` methods.
     """
+    _validate_inputs_custom_constraint(is_valid_fn, transform_fn, reverse_transform_fn)
 
-    def _run(self, function, table_data, reverse=False):
-        table_data = table_data.copy()
-        if self._columns:
-            if reverse:
-                columns = reversed(self._columns)
-            else:
-                columns = self._columns
-
-            for column in columns:
-                try:
-                    table_data = function(table_data, column)
-                except TypeError:
-                    table_data[column] = function(table_data[column])
+    class CustomConstraint(Constraint):
+        """CustomConstraint class.
 
-        else:
-            table_data = function(table_data)
+        Args:
+            transform (callable):
+                Function to replace the ``transform`` method.
+            reverse_transform (callable):
+                Function to replace the ``reverse_transform`` method.
+            is_valid (callable):
+                Function to replace the ``is_valid`` method.
+        """
 
-        return table_data
+        def __init__(self, column_names, **kwargs):
+            self.column_names = column_names
+            self.kwargs = kwargs
 
-    def _run_transform(self, table_data):
-        if self._columns:
-            missing_columns = [col for col in self._columns if col not in table_data.columns]
-            if missing_columns:
-                raise MissingConstraintColumnError(missing_columns=missing_columns)
+        def is_valid(self, data):
+            """Check whether the column values are valid.
 
-        return self._run(self._transform, table_data)
+            Args:
+                table_data (pandas.DataFrame):
+                    Table data.
 
-    def _run_reverse_transform(self, table_data):
-        return self._run(self._reverse_transform, table_data, reverse=True)
+            Returns:
+                pandas.Series:
+                    Whether each row is valid.
+            """
+            valid = is_valid_fn(self.column_names, data, **self.kwargs)
+            if len(valid) != data.shape[0]:
+                raise InvalidFunctionError(
+                    '`is_valid_fn` did not produce exactly 1 True/False value for each row.')
 
-    def _run_is_valid(self, table_data):
-        if self._columns:
-            try:
-                valid = [self._is_valid(table_data, column) for column in self._columns]
-            except TypeError:
-                valid = [self._is_valid(table_data[column]) for column in self._columns]
+            return valid
 
-            return np.logical_and.reduce(valid)
+        def _transform(self, data):
```

Comment:
I thought we were going to override the `transform` method and do something to catch errors so that it is possible to fall back on reject sampling. Currently, the logic only falls back if the constraint raises a `MissingConstraintColumnError`, but that can't happen with a custom constraint. We either need to catch most/all errors and surface it as a `MissingConstraintColumnError`, or provide some special logic to handle that